packages/api/src/external/carequality/document/query-documents.ts

```diff
@@ -0,0 +1,213 @@
+import { Config } from "../../../shared/config";
+import { makeFhirApi } from "../../fhir/api/api-factory";
+import { getAllPages } from "../../fhir/shared/paginated";
+import { isConvertible } from "../../fhir-converter/converter";
+import { appendDocQueryProgress } from "../../../command/medical/patient/append-doc-query-progress";
+import { createCQDocumentQueryRequest } from "./document-query-request";
+import { getOrganizationOrFail } from "../../../command/medical/organization/get-organization";
+import { makeIheGatewayAPI } from "../api";
+import { capture } from "../../../shared/notifications";
+import { MedicalDataSource } from "../../../external";
+import { getCQPatientData } from "../command/cq-patient-data/get-cq-data";
+import { MAPIWebhookStatus } from "../../../command/medical/document/document-webhook";
+import { processPatientDocumentRequest } from "../../../command/medical/document/document-webhook";
+// import { downloadDocs } from "./download-documents";
+
+const s3Utils = new S3Utils(Config.getAWSRegion());
+const s3BucketName = Config.getMedicalDocumentsBucketName();
+const iheGateway = makeIheGatewayAPI();
+
+export const DOCUMENT_QUERY_TIMEOUT = dayjs.duration({ minutes: 2 });
+
+export async function getDocumentsFromCQ({
+  requestId,
+  patient,
+}: {
+  requestId: string;
+  patient: Patient;
+}) {
+  if (!iheGateway) return;
+
+  const { log } = out(`CQ query docs - requestId ${requestId}, M patient ${"123"}`);
+
+  try {
+    const organization = await getOrganizationOrFail({ cxId: patient.cxId });
+    const cqPatientData = await getCQPatientData({ id: patient.id, cxId: patient.cxId });
+
+    const docQueryRequest = createCQDocumentQueryRequest({
+      requestId,
+      cxId: patient.cxId,
+      organization,
+      cqLinks: cqPatientData?.data.links ?? [],
+    });
+
+    await iheGateway.startDocumentsQuery(docQueryRequest);
+
+    // TODO: CHANGE THIS TO THE STRATEGY RAMIL USED: RACE CONTROL
+    const results = await pTimeout(
+      getDocumentQueryResult({ requestId }),
+      DOCUMENT_QUERY_TIMEOUT.asMilliseconds(),
+      "handleDocQueryResponse function timed out!"
+    );
+
+    await handleDocQueryResults(requestId, patient.id, patient.cxId, results, log);
+  } catch (error) {
+    const msg = `Failed to query and process documents in Carequality.`;
+    console.log(`${msg}. Error: ${errorToString(error)}`);
+
+    const updatedPatient = await appendDocQueryProgress({
+      patient: { id: patient.id, cxId: patient.cxId },
+      downloadProgress: { status: "failed" },
+      requestId,
+      source: MedicalDataSource.COMMONWELL,
+    });
+
+    const downloadProgressHasFailed =
+      updatedPatient.data.documentQueryProgress?.download?.status === "failed";
+
+    if (downloadProgressHasFailed) {
+      processPatientDocumentRequest(
+        patient.cxId,
+        patient.id,
+        "medical.document-download",
+        MAPIWebhookStatus.failed
+      );
+    }
```

Comment:
With more than one HIE, I don't think these (CQ/CW) code/logic should be responsible for WH comms - IMO this should be concentrated in Metriport-specific code.

We could look into [events](https://metriport.slack.com/archives/C04DMKE9DME/p1702675771527909) or something else to trigger the logic on the Metriport/domain side.